externallib.php

```diff
@@ -0,0 +1,96 @@
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * External checklist API
+ *
+ * @copyright Davo Smith <moodle@davosmith.co.uk>
+ * @package    mod_checklist
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use mod_checklist\local\checklist_comment_student;
+
+defined('MOODLE_INTERNAL') || die;
+
+require_once("$CFG->libdir/externallib.php");
+require_once("$CFG->dirroot/mod/checklist/locallib.php");
+
+/**
+ * Checklist functions
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mod_checklist_external extends external_api {
+
+    /**
+     * Returns description of method parameters
+     * @return external_function_parameters
+     */
+    public static function update_student_comment_parameters() {
+        return new external_function_parameters(
+            [
+                'comment' => new external_single_structure(
+                    [
+                        'commenttext' => new external_value(PARAM_TEXT, 'content of the comment'),
+                        'checklistitemid' => new external_value(PARAM_INT, 'id of checklist item'),
+                        'cmid' => new external_value(PARAM_INT, 'cmid of checklist module')
+                    ]
+                ),
+            ]
+        );
+    }
+
+    /**
+     * Update or create a student comment for a checklist item.
+     * @param array $params array with comment data.
+     * @return string welcome message
+     */
+    public static function update_student_comment($params): string {
+        global $USER;
+        $params = self::validate_parameters(self::update_student_comment_parameters(), array('comment' => $params));
+        $commentdata = $params['comment'];
+        $cm = get_coursemodule_from_id('checklist', $commentdata['cmid'], 0, false, MUST_EXIST);
+        $context = context_module::instance($cm->id);
+        require_capability('mod/checklist:updateown', $context);
+        require_sesskey();
+        $eventdata = [];
+        $eventdata['context'] = $context;
+        $eventdata['userid'] = $USER->id;
+        $eventdata['objectid'] = $commentdata['checklistitemid'];
+        $eventdata['other'] = ['commenttext' => $commentdata['commenttext']];
+        $existingcomment = checklist_comment_student::get_record([
+            'itemid' => $commentdata['checklistitemid'],
```

Comment:
I can't spot anywhere that you're validating that this checklist item exists and belongs to the checklist with the given 'cmid'. Without that check, a malicious user (who had access to comment on one checklist) could randomly add comments to any checklist item on the site (they would probably not show up in the UI, as they wouldn't be valid for that checklist + user combination, but still not a good idea to let user's arbitrarily add database records).

Reply:
True, I'll make a check for that. Good catch :)